components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt

```diff
@@ -0,0 +1,194 @@
+        overrideHostHeader = builder.overrideHostHeader,
+        responseTimeoutMillis = builder.responseTimeoutMillis,
+        maxHeaderSize = builder.maxHeaderSize,
+        tlsSettings = builder.tlsSettings
+    )
+
+    companion object {
+        const val DEFAULT_RESPONSE_TIMEOUT_MILLIS = 1000
+        const val USE_DEFAULT_MAX_HEADER_SIZE = 0
+        inline fun build(block: Builder.() -> Unit) = Builder().apply(block).build()
+    }
+
+    /**
+     * A builder for [BackendService].
+     */
+    class Builder {
+        var id: Id = Id.GENERIC_APP
+        var path: String = "/"
+        var origins: Set<Origin> = emptySet()
+        var connectionPoolSettings: ConnectionPoolSettings = ConnectionPoolSettings.defaultConnectionPoolSettings()
+        var stickySessionConfig: StickySessionConfig = StickySessionConfig.stickySessionDisabled()
+        var healthCheckConfig: HealthCheckConfig? = null
+        var rewrites: List<RewriteConfig> = emptyList()
+        var overrideHostHeader: Boolean = false
+        var responseTimeoutMillis: Int = DEFAULT_RESPONSE_TIMEOUT_MILLIS
+        var maxHeaderSize: Int = USE_DEFAULT_MAX_HEADER_SIZE
+        var tlsSettings: TlsSettings? = null
+
+        fun build() = BackendService(this)
+    }
+
+    private fun nullIfDisabled(healthCheckConfig: HealthCheckConfig?): HealthCheckConfig? {
+        return if (healthCheckConfig != null && healthCheckConfig.isEnabled) healthCheckConfig else null
+    }
+
+    override fun id(): Id = id
+
+
+    fun idAsString(): String = id.toString()
+
+
+    fun path(): String = path
+
+
+    fun origins(): Set<Origin> = origins
+
+
+    fun connectionPoolConfig(): ConnectionPoolSettings = connectionPoolSettings
+
+
+    fun healthCheckConfig(): HealthCheckConfig? = healthCheckConfig
+
+
+    fun stickySessionConfig(): StickySessionConfig = stickySessionConfig
+
+
+    fun rewrites(): List<RewriteConfig> = rewrites
+
+
+    fun responseTimeoutMillis(): Int = responseTimeoutMillis
+
+
+    fun maxHeaderSize(): Int = maxHeaderSize
+
+
+    fun tlsSettings(): Optional<TlsSettings> = Optional.ofNullable(tlsSettings)
+
+
+    fun isOverrideHostHeader(): Boolean = overrideHostHeader
+
+
+    fun getTlsSettings(): TlsSettings? = tlsSettings().orElse(null)
+
+
+    fun protocol(): Protocol {
+        return if (tlsSettings == null) {
+            Protocol.HTTP
+        } else {
+            Protocol.HTTPS
+        }
+    }
+
+    override fun hashCode(): Int {
+        return Objects.hash(
+            id, path, connectionPoolSettings, origins,
+            healthCheckConfig, stickySessionConfig, rewrites,
+            responseTimeoutMillis, maxHeaderSize
+        )
+    }
+
+    override fun equals(obj: Any?): Boolean {
+        if (this === obj) {
+            return true
+        }
+        if (obj == null || javaClass != obj.javaClass) {
+            return false
+        }
+        val other = obj as BackendService
+        return (id == other.id
```

Comment:
i think this might be the same as the generated equals method on the data class. Pretty sure it compares each property, but maybe reviewing the generated implementation will tell us whether we need to override this (and others like hashCode) at all.

Reply:
For toString it seems like the choice of curl braces vs parenthesis is the difference.